### DISCUSSION TITLE
Adding `check` parameters as a shortcut for raise_for_status

### DISCUSSION THREAD

**Discussion by mik-laj** [encode/httpx#1737] [Ideas]
Hi.

I would like to be able to send an HTTP request, check response error codes, and read the results with a single expression. Currently, for this purpose, I most often create utilities function:
```python
def api_request(method, path, base_url="http://localhost:8080/api/v1/", **kwargs) -> Dict:
    response = httpx.request(
        method=method,
        url=f"{base_url}/{path}",
        **kwargs,
    )
    response.raise_for_status()
    return response.json()
```
I think, that this is such a common pattern that it would be worth simplifying it in the library itself.
To do this, we can add another check parameter, which will cause the `raise_for_status` function to be called before the result is returned. This will allow us to download a simple JSON file in one line while still handling exceptional situations.
```python
api_object = requests.get('http://....', check=True).json()
```
This will be in line with the assumptions of the standard library - `subprocess.run`, so it will make using this library even easier. Developers will not have to learn about an additional method in the inner class.
```python
subprocess.run(["external-tool"], check=True)
```
Best regards,
Kamil Breguła

**Comment by lovelydinosaur** [comment]
At *this point in time* it'd be a "no" from me. I don't *really* want us to consider extending any API past the boundaries that parity-with-requests gives us, until we're past 1.0. (Unless there's some *very* compelling reason to do so.)

Note that you *can* already do the following...

```python
r = httpx.get('http://....').raise_for_status().json()
```

Which is functionally pretty much the same example use-case you've given, but just a slightly different syntax to achieve it.

I can *imagine* that there *might* be something here for us to explore in the future. For example, I can see something like the following might fit in well with our existing API...

```python
client = httpx.Client(raise_for_status=True)  # A client instance that automatically calls `raise_for_status()` on responses.
r = client.get('http://....').json()
```

**Reply by mik-laj** [reply]
> Note that you _can_ already do the following...
> 
> ```python
> r = httpx.get('http://....').raise_for_status().json()
> ```

I'm afraid that won't work because `Response.raise_for_status` returns `None` and not `Request`.
https://github.com/encode/httpx/blob/47d712c01cdde6574b2a94e43a94e5ae6e4f2faf/httpx/_models.py#L1384

**Reply by lovelydinosaur** [reply]
Oh right. We *might* want to consider having it returning the response instance, which is what `requests` does. Not sure.
